File: app/src/main/java/com/ywwynm/everythingdone/fragments/LongTextDialogFragment.java
package com.ywwynm.everythingdone.fragments;

import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.view.Window;
import android.view.WindowManager;
import android.widget.ScrollView;
import android.widget.TextView;

import com.ywwynm.everythingdone.R;
import com.ywwynm.everythingdone.utils.DisplayUtil;
import com.ywwynm.everythingdone.utils.EdgeEffectUtil;

/**
 * Created by ywwynm on 2016/4/23.
 * long text dialog fragment
 */
public class LongTextDialogFragment extends BaseDialogFragment {

    public static final String TAG = "LongTextDialogFragment";

    private int mAccentColor;

    private String mTitle;
    private String mContent;
    private String mConfirmText = null;

    private boolean mShowCancel = false;

    private View.OnClickListener mConfirmListener;
    private View.OnClickListener mCancelListener;

    private boolean mConfirmed = false;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        super.onCreateView(inflater, container, savedInstanceState);

        TextView tvTitle       = f(R.id.tv_title_long_text);
        TextView tvContent     = f(R.id.tv_content_long_text);
        TextView tvCancelAsBt  = f(R.id.tv_cancel_as_bt_long_text);
        TextView tvConfirmAsBt = f(R.id.tv_confirm_as_bt_long_text);

        if (mTitle != null) {
            tvTitle.setTextColor(mAccentColor);
            tvTitle.setText(mTitle);
        } else {
            tvTitle.setVisibility(View.GONE);
        }

        if (mContent != null) {
            tvContent.setText(mContent);
        } else {
            tvContent.setVisibility(View.GONE);
        }

        if (mShowCancel) {
            tvCancelAsBt.setVisibility(View.VISIBLE);
            tvCancelAsBt.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    dismiss();
                }
            });
        }

        if (mConfirmText != null) {
            tvConfirmAsBt.setText(mConfirmText);
        }
        tvConfirmAsBt.setTextColor(mAccentColor);
        tvConfirmAsBt.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mConfirmListener != null) {
                    mConfirmListener.onClick(v);
                }
                mConfirmed = true;
                dismiss();
            }
        });

        final View v1 = f(R.id.view_separator_1);
        final View v2 = f(R.id.view_separator_2);
        final ScrollView sv = f(R.id.sv_long_text);
        sv.getViewTreeObserver().addOnScrollChangedListener(
                new ViewTreeObserver.OnScrollChangedListener() {
                    @Override
                    public void onScrollChanged() {
                        int scrollY = sv.getScrollY();
                        if (scrollY <= 0) {
                            v1.setVisibility(View.INVISIBLE);
                            v2.setVisibility(View.VISIBLE);
                        } else if (scrollY >= sv.getChildAt(0).getHeight() - sv.getHeight()) {
                            v1.setVisibility(View.VISIBLE);
                            v2.setVisibility(View.INVISIBLE);
                        } else {
                            v1.setVisibility(View.VISIBLE);
                            v2.setVisibility(View.VISIBLE);
                        }
                    }
                }
        );
        EdgeEffectUtil.forScrollView(sv, mAccentColor);

        return mContentView;
    }

    @Override
    protected int getLayoutResource() {
        return R.layout.fragment_long_text;
    }

    @Override
    public void onResume() {
        super.onResume();
        float screenDensity = DisplayUtil.getScreenDensity(getActivity());
        Window window = getDialog().getWindow();
        if (window != null) {
            window.setLayout((int) (screenDensity * 320), WindowManager.LayoutParams.WRAP_CONTENT);
        }
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        if (!mConfirmed && mCancelListener != null) {
            mCancelListener.onClick(f(R.id.tv_cancel_as_bt_long_text));
        }
        mConfirmListener = null;
        mCancelListener  = null;
        super.onDismiss(dialog);
    }

    public void setAccentColor(int accentColor) {
        mAccentColor = accentColor;
    }

    public void setTitle(String title) {
        mTitle = title;
    }

    public void setContent(String content) {
        mContent = content;
    }

    public void setConfirmText(String confirmText) {
        mConfirmText = confirmText;
    }

    public void setShowCancel(boolean showCancel) {
        mShowCancel = showCancel;
    }

    public void setConfirmListener(View.OnClickListener confirmListener) {
        mConfirmListener = confirmListener;
    }

    public void setCancelListener(View.OnClickListener cancelListener) {
        mCancelListener = cancelListener;
    }
}

File: app/src/main/java/com/ywwynm/everythingdone/fragments/LongTextDialogFragment.kt
@file:Suppress("DEPRECATION", "OVERRIDE_DEPRECATION")

package com.ywwynm.everythingdone.fragments

import android.content.DialogInterface
import android.os.Bundle
import androidx.core.content.ContextCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowManager
import android.widget.ScrollView
import android.widget.TextView

import com.ywwynm.everythingdone.R
import com.ywwynm.everythingdone.model.ThingBackground
import com.ywwynm.everythingdone.utils.BackgroundUtil
import com.ywwynm.everythingdone.utils.DisplayUtil
import com.ywwynm.everythingdone.utils.EdgeEffectUtil

/**
 * Created by ywwynm on 2016/4/23.
 * Translated to Kotlin by ywwynm and Claude Opus 4.7 on 2026/5/20.
 * long text dialog fragment
 */
open class LongTextDialogFragment : BaseDialogFragment() {

    private var mAccentColor: Int = 0
    /** Phase 8: full ThingBackground for gradient title / confirm. */
    private var mAccentBackground: ThingBackground? = null

    private var mTitle: String? = null
    private var mContent: String? = null
    private var mConfirmText: String? = null

    private var mShowCancel: Boolean = false

    private var mConfirmListener: View.OnClickListener? = null
    private var mCancelListener: View.OnClickListener? = null

    private var mConfirmed: Boolean = false

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        super.onCreateView(inflater, container, savedInstanceState)

        if (mAccentColor == 0 && mAccentBackground == null) {
            mAccentColor = ContextCompat.getColor(
                activity!!, R.color.app_chrome_on_surface_strong
            )
        }

        val tvTitle: TextView       = f(R.id.tv_title_long_text)!!
        val tvContent: TextView     = f(R.id.tv_content_long_text)!!
        val tvCancelAsBt: TextView  = f(R.id.tv_cancel_as_bt_long_text)!!
        val tvConfirmAsBt: TextView = f(R.id.tv_confirm_as_bt_long_text)!!

        if (mTitle != null) {
            tvTitle.text = mTitle
            if (mAccentBackground != null) {
                BackgroundUtil.applyTextBackground(tvTitle, mAccentBackground)
            } else {
                tvTitle.setTextColor(mAccentColor)
            }
        } else {
            tvTitle.visibility = View.GONE
        }

        if (mContent != null) {
            tvContent.text = mContent
        } else {
            tvContent.visibility = View.GONE
        }

        if (mShowCancel) {
            tvCancelAsBt.visibility = View.VISIBLE
            tvCancelAsBt.setOnClickListener { dismiss() }
        }

        if (mConfirmText != null) {
            tvConfirmAsBt.text = mConfirmText
        }
        if (mAccentBackground != null) {
            BackgroundUtil.applyTextBackground(tvConfirmAsBt, mAccentBackground)
        } else {
            tvConfirmAsBt.setTextColor(mAccentColor)
        }
        tvConfirmAsBt.setOnClickListener { v ->
            if (mConfirmListener != null) {
                mConfirmListener!!.onClick(v)
            }
            mConfirmed = true
            dismiss()
        }

        val v1: View = f(R.id.view_separator_1)!!
        val v2: View = f(R.id.view_separator_2)!!
        val sv: ScrollView = f(R.id.sv_long_text)!!
        sv.viewTreeObserver.addOnScrollChangedListener {
            val scrollY = sv.scrollY
            if (scrollY <= 0) {
                v1.visibility = View.INVISIBLE
                v2.visibility = View.VISIBLE
            } else if (scrollY >= sv.getChildAt(0).height - sv.height) {
                v1.visibility = View.VISIBLE
                v2.visibility = View.INVISIBLE
            } else {
                v1.visibility = View.VISIBLE
                v2.visibility = View.VISIBLE
            }
        }
        EdgeEffectUtil.forScrollView(sv, mAccentColor)

        return mContentView
    }

    override fun getLayoutResource(): Int = R.layout.fragment_long_text

    override fun onResume() {
        super.onResume()
        val screenDensity: Float = DisplayUtil.getScreenDensity(activity)
        val window: Window? = dialog!!.window
        window?.setLayout((screenDensity * 320).toInt(), WindowManager.LayoutParams.WRAP_CONTENT)
    }

    override fun onDismiss(dialog: DialogInterface) {
        if (!mConfirmed && mCancelListener != null) {
            mCancelListener!!.onClick(f(R.id.tv_cancel_as_bt_long_text))
        }
        mConfirmListener = null
        mCancelListener  = null
        super.onDismiss(dialog)
    }

    open fun setAccentColor(accentColor: Int) {
        mAccentColor = accentColor
        mAccentBackground = null
    }

    /** Phase 8: gradient-aware accent. */
    open fun setAccentBackground(bg: ThingBackground?) {
        mAccentBackground = bg
        if (bg != null) mAccentColor = bg.color
    }

    open fun setTitle(title: String?) {
        mTitle = title
    }

    open fun setContent(content: String?) {
        mContent = content
    }

    open fun setConfirmText(confirmText: String?) {
        mConfirmText = confirmText
    }

    open fun setShowCancel(showCancel: Boolean) {
        mShowCancel = showCancel
    }

    open fun setConfirmListener(confirmListener: View.OnClickListener?) {
        mConfirmListener = confirmListener
    }

    open fun setCancelListener(cancelListener: View.OnClickListener?) {
        mCancelListener = cancelListener
    }

    companion object {
        const val TAG: String = "LongTextDialogFragment"
    }
}
